### PR TITLE
Fix TLS 1.3 issues on JDKs which support it

### PR DESCRIPTION
**Pull request by jrudolph**
Refs #29110, #29111

Commits have detailed information about the changes.

(Please don't squash!)

### CODE REVIEW

**Comment by jrudolph** [comment]
Tests need to be run on JDK 11+ to see TLS v1.3 tests.

File: akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala [hunk 1]
```diff
@@ -102,469 +102,486 @@ class TlsSpec extends StreamSpec(TlsSpec.configOverrides) with WithLogCapturing
...
-    val sslContext = initSslContext()
+    if (JavaVersion.majorVersion >= 11)
+      "work for TLSv1.3" must { workFor("TLSv1.3", TLS13Ciphers) }
 
-    val debug = Flow[SslTlsInbound].map { x =>
-      x match {
-        case SessionTruncated   => system.log.debug(s" ----------- truncated ")
-        case SessionBytes(_, b) => system.log.debug(s" ----------- (${b.size}) ${b.take(32).utf8String}")
+    def workFor(protocol: String, ciphers: Set[String]): Unit = {
+      val sslContext = initSslContext(protocol)
```
Comment (jrudolph) [inline akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala:111]: You can hide whitespace changes for this file when reviewing, mostly only indentation changed.

**Review by raboof — APPROVED** [review]
Pretty intricate, but convincing!

(let's indeed not squash this when merging)

File: akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala [hunk 2]
```diff
@@ -102,469 +102,486 @@ class TlsSpec extends StreamSpec(TlsSpec.configOverrides) with WithLogCapturing
...
-      val scenario = SingleBytes
+      "reliably cancel subscriptions when TransportIn fails early" in assertAllStagesStopped {
+        val ex = new Exception("hello")
+        val (sub, out1, out2) =
+          RunnableGraph
+            .fromGraph(
+              GraphDSL.create(Source.asSubscriber[SslTlsOutbound], Sink.head[ByteString], Sink.head[SslTlsInbound])(
+                (_, _, _)) { implicit b => (s, o1, o2) =>
+                val tls = b.add(clientTls(EagerClose))
+                s ~> tls.in1;
```
Comment (ennru) [inline akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala:480]:
```suggestion
                s ~> tls.in1
```

File: akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala [hunk 3]
```diff
@@ -102,469 +102,486 @@ class TlsSpec extends StreamSpec(TlsSpec.configOverrides) with WithLogCapturing
...
+                tls.out1 ~> o1
+                o2 <~ tls.out2;
```
Comment (ennru) [inline akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala:482]:
```suggestion
                o2 <~ tls.out2
```

File: akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala [hunk 4]
```diff
@@ -102,469 +102,486 @@ class TlsSpec extends StreamSpec(TlsSpec.configOverrides) with WithLogCapturing
...
-          .via(debug)
+      "reliably cancel subscriptions when UserIn fails early" in assertAllStagesStopped {
+        val ex = new Exception("hello")
+        val (sub, out1, out2) =
+          RunnableGraph
+            .fromGraph(
+              GraphDSL.create(Source.asSubscriber[ByteString], Sink.head[ByteString], Sink.head[SslTlsInbound])(
+                (_, _, _)) { implicit b => (s, o1, o2) =>
+                val tls = b.add(clientTls(EagerClose))
+                Source.failed[SslTlsOutbound](ex) ~> tls.in1;
```
Comment (ennru) [inline akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala:503]:
```suggestion
                Source.failed[SslTlsOutbound](ex) ~> tls.in1
```

File: akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala [hunk 5]
```diff
@@ -102,469 +102,486 @@ class TlsSpec extends StreamSpec(TlsSpec.configOverrides) with WithLogCapturing
...
+                tls.out1 ~> o1
+                o2 <~ tls.out2;
```
Comment (ennru) [inline akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala:505]:
```suggestion
                o2 <~ tls.out2
```

**Review by ennru — APPROVED** [review]
Should it be documented that half-open TLS connections are not supported?

**Comment by raboof** [comment]
> Should it be documented that half-open TLS connections are not supported?

That might be good. We still support half-open when incoming is closed, just not half-open when outgoing is closed, right? Or did I mix things up?

**Comment by jrudolph** [comment]
> > Should it be documented that half-open TLS connections are not supported?

I amended the docs on TLS which had a large paragraph about this very issue.

> That might be good. We still support half-open when incoming is closed, just not half-open when outgoing is closed, right? Or did I mix things up?

Depends on which `incoming` you mean. We don't support half-closed on the encrypted at all (or if, only accidentally). On the plaintext side, the support configuration via `TLSClosing`.

**Review by raboof — APPROVED** [review]
:+1: added #29141 to track implementing half-closing for TLS 1.3

**Comment by jrudolph** [comment]
> +1 added #29141 to track implementing half-closing for TLS 1.3

:+1: Thanks!

**Pull request merged**